geo/geomfn: Implements ST_LongestLine

abhishek20123g opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/48978

This PR implements ST_LongestLine({geometry, geometry}) builtin
function. Which allow us to find the LineString corresponds to
the maximum distance between given two geometrical objects.

Release note (sql change): This PR implements ST_LongestLine({geometry,
geometry}) builtin function.

abhishek20123g committed:
geo/geomfn: Implements ST_LongestLine

Fixes https://github.com/cockroachdb/cockroach/issues/48978

This PR implements ST_LongestLine({geometry, geometry}) builtin
function. Which allow us to find the LineString corresponds to
the maximum distance between given two geometrical objects.

Release note (sql change): This PR implements ST_LongestLine({geometry,
geometry}) builtin function.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50018)
<!-- Reviewable:end -->

abhishek20123g commented:
Small Overview
* Modified `DistanceUpdater` interface as there seems a requirement to know the order of geometrical objects. Added extra function `TwistGeometries()`(`PostGIS` use `twist` name integer variable to preserve the order) which may require to handle the functionality during twisting.
https://github.com/postgis/postgis/blob/54399b9f6b0f02e8db9444f9f042b8d4ca6d4fa4/liblwgeom/measures.h#L55
* Added extra test-cases for testing edge cases corresponding to twisting of geometrical objects, Thought there we may require to add extra that may be verified by `ST_ShortestLine` as `ST_LongestLine` do not covers that part of code.
* Divided `LongestLineString` into two-part as that may be helpful in future for `ST_ShortestLine`

otan commented on pkg/geo/geogfn/distance.go:
```diff
@@ -291,6 +291,9 @@ func (u *geographyMinDistanceUpdater) IsMaxDistance() bool {
 }
 
+// TwistGeometries implements the geodist.DistanceUpdater interface.
+func (u *geographyMinDistanceUpdater) TwistGeometries() {}
```
can you panic here with unimplemented?

otan commented on pkg/geo/geomfn/distance.go:
```diff
@@ -20,6 +20,19 @@ import (
+const (
+	// geometricalObjectsTwisted represents that the given order of two geometrical
+	// objects has been twisted.
+	geometricalObjectsTwisted geometricalObjectsOrder = -1
```
I think i prefer the name Flipped. Twisted doesn't sound right.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -1610,6 +1610,30 @@ Note If the result has zero or one points, it will be returned as a POINT. If it
+					`given geometries. 
+
+Note if geometries are the same, it will return the LineString with the maximum distance between the geometry's ` +
+					`vertexes. The function will return the first longest line if more than one is found.`,
```
can you replace the phrase "first longest line" with "one of the longest lines"? i think the term "first" is confusing. alternatively, "the longest line that was discovered first when comparing maximum distances"

otan commented on pkg/sql/logictest/testdata/logic_test/geospatial:
```diff
@@ -504,116 +504,117 @@ Square (right)                            POINT (0.5 0.5)
   ST_Distance(a.geom, b.geom),
-  ST_MaxDistance(a.geom, b.geom)
+  ST_MaxDistance(a.geom, b.geom),
+  ST_AsText(ST_LongestLine(a.geom, b.geom))
```
can you separate this out into a separate select statement?

otan commented on pkg/geo/geodist/geodist.go:
```diff
@@ -259,11 +270,16 @@ func onShapeEdgesToShapeEdges(c DistanceCalculator, a shapeWithEdges, b shapeWit
+				c.DistanceUpdater().TwistGeometries()
+				if projectVertexToEdge(c, bEdge.V0, aEdge) ||
 					projectVertexToEdge(c, bEdge.V1, aEdge) {
 					return true
```
do we need to twist again before returning true?

otan commented on pkg/geo/geomfn/distance.go:
```diff
@@ -333,6 +380,9 @@ func (u *geomMinDistanceUpdater) IsMaxDistance() bool {
 }
 
+// TwistGeometries implements the geodist.DistanceUpdater interface.
+func (u *geomMinDistanceUpdater) TwistGeometries() {}
```
panic here with unimplemented.

abhishek20123g replied:
Sure,
If I understand it right we have to write a panic message like,
```
panic("Unimplemented ...")
```
wouldn't it make `ST_distance`, `MinDistance` inexecutable for a while?

otan replied:
oh, you use this for all cases, my bad.
can you add a comment that this is unimplemented pending `ST_ShortestLine` / #49035 for geomfn, and for geogfn adding a  comment that this is unimplemented as we do not care about ordering.

otan reviewed (commented):
Looking good! Nit with the name `Twist` and a few places to add panics.

otan reviewed (commented):
Thanks for the updates - one little thing to make this easier to read, which didn't seem addressed from earlier.
